### PR TITLE
Add benchmarks for watch over websocket and http

### PR DESCRIPTION
**Pull request by smarterclayton**
```
$ godep go test ./pkg/apiserver -benchmem -run=XXX -bench=BenchmarkWatch
PASS
BenchmarkWatchHTTP-8           20000         97169 ns/op       15126 B/op        197 allocs/op
BenchmarkWatchWebsocket-8      10000        102608 ns/op       18509 B/op        205 allocs/op
```

@kubernetes/sig-api-machinery

### CODE REVIEW

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
This is just to provide a baseline for measurement with future refactors

**Comment by smarterclayton** [comment]
@k8s-bot unit test this

File: pkg/apiserver/watch_test.go [hunk 1]
```diff
@@ -396,3 +400,101 @@ func TestWatchHTTPTimeout(t *testing.T) {
...
+
+	actions := []watch.EventType{watch.Added, watch.Modified, watch.Deleted}
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		simpleStorage.fakeWatch.Action(actions[i%len(actions)], &items[i%len(items)])
+	}
+	simpleStorage.fakeWatch.Stop()
+	wg.Wait()
+	b.StopTimer()
```
Comment (nikhiljindal) [inline pkg/apiserver/watch_test.go]: We dont print the time or check that the time should be less than X mins?

Comment (smarterclayton) [inline pkg/apiserver/watch_test.go]: Benchmarks in Go don't typically do that - this is mostly for measurement of cost and overhead.  We don't run benchmarks during normal code paths.

Comment (smarterclayton) [inline pkg/apiserver/watch_test.go]: Times and numbers are handled by the go test framework (see the pasted command I provided)

**Comment by smarterclayton** [comment]
Any other comments?

**Comment by smarterclayton** [comment]
@liggitt review if you have a chance

**Comment by smarterclayton** [comment]
This doesn't add code that a) can flake or b) is called in normal path.  Self applying LGTM so I can run comparisons.

**Comment by smarterclayton** [comment]
ok to test

File: pkg/apiserver/watch_test.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/apiserver/watch_test.go]: Might be a good idea to mention that command as a comment in code

**Comment by nikhiljindal** [comment]
LGTM, thanks

File: pkg/apiserver/watch_test.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/apiserver/watch_test.go]: It probably should be added to the general developer docs.  I'll add it as a follow up.

**Comment by smarterclayton** [comment]
@k8s-bot test this

**Comment by smarterclayton** [comment]
There are lies, damn lies, and benchmarks.  I realized that the groups the apiserver test were binding api.Pod to both group versions, so that triggered ConvertToVersion to do a default reflect conversion, so the numbers were way off.  Much closer now, also added doc to the PR.

**Comment by smarterclayton** [comment]
From the protobuf branch:

```
BenchmarkWatchHTTP-8     	   10000	    100317 ns/op	   12911 B/op	     210 allocs/op
BenchmarkWatchWebsocket-8	   20000	     98927 ns/op	   13229 B/op	     217 allocs/op
BenchmarkWatchProtobuf-8 	   30000	     52011 ns/op	   12504 B/op	     164 allocs/op
```

At that point the protobuf code is dominated by conversion (~90% of the total CPU and the vast majority of allocations).  The `reflect.Value.call` shows up in profiles as creating most of the objects - I suspect we can get a big win on conversion by generating a type switch instead of reflecting into the method calls.

**Pull request merged**